### PR TITLE
Populate all know FQDN to create  LE certificate

### PR DESCRIPTION
**Pull request by stephdl**
The goal of the PR

https://github.com/NethServer/dev/issues/6353

Click the button then take

    Fqdn of the server (SystemName.DomainName)
    Domain of the server (DomainName)
    All hosts with type self of hosts esmith database
    All domains from domains esmith database

Sort the array of domain duplicates name and create the text inputs

Add validators to determine what is the error

Screenshot (21)
Screenshot (20)
Screenshot (19)

### CODE REVIEW

File: api/system-certificate/read [hunk 1]
```diff
@@ -61,6 +65,24 @@ if ($cmd eq 'list') {
...
+    my @domains = ();
+
+    foreach my $d ($dbd->get_all()) {
+        my $domain = $d->key;
+        push @domains, $domain;
+    }
+    foreach my $h ($dbh->get_all_by_prop('type'=>'self')) {
+        my $domain = $h->key;
+        push @domains, $domain;
+    }
```
Comment (gsanchietti) [inline api/system-certificate/read:79]: I'm not sure that mail domains should be added to LE: usually the mail domain is not a FQDN.

Let's get a real example: `nethesis.it`.
The MX record points to `nethservice.nethesis.it`, which is the FQDN of the mail server.

File: api/system-certificate/read [hunk 2]
```diff
@@ -61,6 +65,24 @@ if ($cmd eq 'list') {
...
+
+    push @domains, $db->get('SystemName').'.'.$db->get('DomainName'),$db->get('DomainName');
+    my $domains = join (',',do { my %seen; grep { !$seen{$_}++ } @domains });
+
+    $out->{'KnowDomains'} = $domains;
```
Comment (gsanchietti) [inline api/system-certificate/read:84]: Please reword it to a more correct form:
```suggestion
    $out->{'KnownDomains'} = $domains;
```

File: ui/public/i18n/language.json [hunk 3]
```diff
@@ -390,7 +390,8 @@
     "cert_encrypt_ok": "Let's encrypt requested successfully!",
     "cert_encrypt_error": "Let's encrypt not requested.",
     "cert_self_ok": "Self-signed certificate edited successfully!",
-    "cert_self_error": "Self-signed certificate not edited."
+    "cert_self_error": "Self-signed certificate not edited.",
+    "add_Knowdomain": "Add know FQDN"
```
Comment (gsanchietti) [inline ui/public/i18n/language.json:394]:
```suggestion
    "add_Knowdomain": "Add all known FQDN"
```

File: ui/src/components/system/Certificates.vue [hunk 4]
```diff
@@ -1184,6 +1198,19 @@ export default {
         }
       );
     },
+    addKnowDomains() {
+
+      const existingDomain = 
+        this.letsEncryptCertificate.LetsEncryptDomains.filter(function( obj ) {
+          return obj.name !== '';
+        });;
+      const knowDomain = this.letsEncryptCertificate.KnowDomains;
```
Comment (gsanchietti) [inline ui/src/components/system/Certificates.vue:1207]: Use 'known` instead of `know`

**Review by gsanchietti — CHANGES_REQUESTED** [review]
The whole idea is good, there a still some few things to fix.

Also, it seems to me that we sill miss the parsing of the validator script.

File: api/system-certificate/read (hunk 1 above)
Comment (stephdl) [reply api/system-certificate/read:74]: done

File: api/system-certificate/read (hunk 2 above)
Comment (stephdl) [reply api/system-certificate/read]: done

File: ui/public/i18n/language.json (hunk 3 above)
Comment (stephdl) [reply ui/public/i18n/language.json]: done

File: ui/src/components/system/Certificates.vue (hunk 4 above)
Comment (stephdl) [reply ui/src/components/system/Certificates.vue]: done

File: ui/public/i18n/language.json [hunk 5]
```diff
@@ -378,7 +378,7 @@
     "renew_days": "Renew days",
     "request": "Request",
     "cockpit_reload_warning_title": "Notice.",
-    "cockpit_reload_warning_message": "The new certificate is applied to the Cockpit Server Manager 90 seconds after the last user disconnects.",
+    "cockpit_reload_warning_message": "The new certificate is applied to the Cockpit Server Manager 90 seconds after the last user disconnects. You cannot ask more of 10 certificate requests by 3 hours, the error of validation are included in the rate limit.",
```
Comment (gsanchietti) [inline ui/public/i18n/language.json:381]: > You cannot ask more of 10 certificate requests by 3 hours, the error of validation are included in the rate limit.

This is not really true, it applies only to new account so in our case it's a limit or the validator. See https://letsencrypt.org/docs/rate-limits/

I'd remove the additional because it's a little out of context.

File: api/system-certificate/validate [hunk 6]
```diff
@@ -64,9 +64,25 @@ function validate_letsencrypt($props)
 
     $cmd = "/usr/libexec/nethserver/letsencrypt-certs -e '".escapeshellarg($props['LetsEncryptMail'])."' -t -d ". escapeshellarg(join(",",$domains))." 2>&1";
     exec($cmd, $output, $ret);
```
[the remark this answers is not in the document]
Comment (stephdl) [reply api/system-certificate/validate:66]: done

File: api/system-certificate/validate [hunk 7]
```diff
@@ -64,9 +64,25 @@ function validate_letsencrypt($props)
...
+
     if ($ret!=0) {
-        $v->addValidationError('LetsEncryptDomains', "letencrypt_domain_failed");
-        error($v);
+        $FailedDomain = array();
+        if (preg_grep('/Challenge failed/',$output)) {
+            foreach ($output as $line) {
+                if (preg_match('/Challenge failed for domain/',$line)) {
```
[the remark this answers is not in the document]
Comment (stephdl) [reply api/system-certificate/validate]: done

File: api/system-certificate/validate [hunk 8]
```diff
@@ -75,6 +91,7 @@ function validate_letsencrypt($props)
     }
 }
 
+
```
[the remark this answers is not in the document]
Comment (stephdl) [reply api/system-certificate/validate]: done

File: ui/public/i18n/language.json (hunk 5 above)
Comment (stephdl) [reply ui/public/i18n/language.json]: done

File: ui/public/i18n/language.json [hunk 9]
```diff
@@ -1217,7 +1217,10 @@
     "TokenValidationError": "Verification code is not valid.",
     "TokenValidationOK": "Verification code is valid!",
     "valid_hostname_0_9223372036854775807":"Invalid domain name, it must be something like 'domain.org'",
-    "Host_already_in_use_by_another_smarthost_relay": "This smart host is already used for another smart host relay, you cannot use it as default for the server"
+    "Host_already_in_use_by_another_smarthost_relay": "This smart host is already used for another smart host relay, you cannot use it as default for the server",
+    "letsencrypt_registration_failed_rate_limit_reached": "Rate limit reached. Please wait at least 3 hours before performing a new request or make sure to change your public IP address",
+    "letsencrypt_unknow_error":"Challenge failed, please see the error message",
```
Comment (gsanchietti) [inline ui/public/i18n/language.json:1222]:
```suggestion
    "letsencrypt_unknow_error":"Challenge failed, please see the error message below.",
```

File: ui/public/i18n/language.json [hunk 10]
```diff
@@ -390,7 +390,8 @@
     "cert_encrypt_ok": "Let's encrypt requested successfully!",
     "cert_encrypt_error": "Let's encrypt not requested.",
     "cert_self_ok": "Self-signed certificate edited successfully!",
-    "cert_self_error": "Self-signed certificate not edited."
+    "cert_self_error": "Self-signed certificate not edited.",
+    "add_Knowndomain": "Add all known FQDN"
```
Comment (gsanchietti) [inline ui/public/i18n/language.json:394]:
```suggestion
    "add_Knowndomain": "Add all known FQDNs"
```

File: ui/src/components/system/Certificates.vue [hunk 11]
```diff
@@ -483,7 +483,15 @@
...
-
+              <div class="form-group">
+                <div class="col-sm-3 control-label"></div>
+                <div class="col-sm-9">
+                  <button @click="addKnownDomains()" class="btn btn-default" type="button">
+                    <span class="fa fa-plus card-icon-def"></span>
+                    {{$t('certificates.add_Knowndomain')}}
+                  </button>
+                </div>
+              </div>
```
Comment (gsanchietti) [inline ui/src/components/system/Certificates.vue:494]: I've had brief chat also with Edoardo: having this button above the form is a bit misleading.

Please move the button below the form, just on the right of the existing "Add domain" button.

**Review by gsanchietti — CHANGES_REQUESTED** [review]

File: ui/src/components/system/Certificates.vue (hunk 11 above)
Comment (stephdl) [reply ui/src/components/system/Certificates.vue:494]: Screenshot (23)

**Review by gsanchietti — APPROVED** [review]
:clap:
